Prints the packages that will be uploaded when using the --verbose option

VikramJayanthi17 opened the pull request:
As per discussion in #381, twine now prints the names of the packages that it will attempt to upload when the `--verbose` option is selected. The example output below is from the terminal on my MacOs(Mojave) machine. 

<img width="1677" alt="Screen Shot 2020-06-05 at 2 44 11 PM" src="https://user-images.githubusercontent.com/35483161/83925342-c3510f80-a73b-11ea-8075-7c1823fe095d.png">

This change was suggested by @bhrutledge as a simple first change to the `--verbose` option using the existing verbose pattern before more comprehensive changes are implemented using the `stdlib` logging system.

sigmavirus24 commented:
Please fix your failing tests https://travis-ci.org/github/pypa/twine/builds/695224649 and clean up your commit history on the branch

bhrutledge commented on twine/commands/upload.py:
```diff
@@ -64,6 +64,11 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
 
+    if upload_settings.verbose:
+        print("\nList of packages that will be uploaded:")
+        for package in uploads:
```
I think this is the source of the failing type checks (see the [Travis log](https://travis-ci.org/github/pypa/twine/jobs/695272199#L242-L251)), because this results in mypy inferring that `package` is a `str`, but later it's assigned a `PacakgeFile`  instance. You can resolve it by using a different variable name, e.g. `upload_filename`.

bhrutledge commented on twine/commands/upload.py:
```diff
@@ -64,6 +64,11 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
+        print("\nList of packages that will be uploaded:")
+        for package in uploads:
+            print("  " + os.path.basename(package))
+        print("\n")
```
Seeing your screenshot makes me think this could be more useful, e.g.:

```
$ twine upload -r testpypi --verbose dist/* 
Uploading distributions to https://test.pypi.org/legacy/
  dist/example_pkg_bhrutledge-0.0.4a4-py3-none-any.whl (2.6 KB)
  dist/example-pkg-bhrutledge-0.0.4a3.tar.gz (1.3 KB)
  dist/example-pkg-bhrutledge-0.0.4a4.tar.gz (1.4 KB)

Enter your username: 
```

Specifically:

- Paired with the "Uploading distributions" line
- Showing the full path, instead of the file name
- Showing the file size (just an idea, could be done later in a separate PR)

sigmavirus24 replied:
> Showing the file size (just an idea, could be done later in a separate PR)

I like this. I wonder if we can also show that we're uploading the associated `.asc` file too so people stop asking why we aren't uploading it (when we are)

bhrutledge replied:
> I wonder if we can also show that we're uploading the associated `.asc` file too so people stop asking why we aren't uploading it (when we are)

I wondered the same thing, and suggested it in https://github.com/pypa/twine/issues/381#issuecomment-538103137. However, I thought that might be better handled in a separate PR, since it's a separate operation, later in `upload`:

https://github.com/pypa/twine/blob/31cad981425b2ef8ee17f74925320acc356ab3a6/twine/commands/upload.py#L82-L86

See this makes me wonder if `add_gpg_signature` would be the right spot for a verbose message.

VikramJayanthi17 replied:
> Please fix your failing tests https://travis-ci.org/github/pypa/twine/builds/695224649 and clean up your commit history on the branch
> Paired with the "Uploading distributions" line
Showing the full path, instead of just the file name

Thanks for help! I will fix the tests, and make these changes in a PR. 

> Showing the file size (just an idea, could be done later in a separate PR)
> we can also show that we're uploading the associated .asc file too so people stop asking why we aren't uploading it (when we are)

I will create a checklist on #381 that includes these changes as well as the others discussed in #381.

bhrutledge replied:
Sounds good, @VikramJayanthi17. If you're game, I think it would be okay to show the file size as part of this PR.

VikramJayanthi17 replied:
> If you're game, I think it would be okay to show the file size as part of this PR.

Sounds good, I will add it to this PR.

sigmavirus24 replied:
> See this makes me wonder if `add_gpg_signature` would be the right spot for a verbose message.

This makes me think we're printing this way too early if we can't rely on the PackageFile instance for each item. We can print the filename and the asc filename at that point

bhrutledge replied:
@sigmavirus24 Can we hold on the signature discussion for now? It'll be implemented in a subsequent PR.

sigmavirus24 replied:
@bhrutledge no, because the signatures should be printed in the same place and so that makes me believe that this is the wrong place to be printing this informatoin.

sigmavirus24 commented on twine/commands/upload.py:
```diff
@@ -60,6 +60,12 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
+        for dist_to_upload in uploads:
+            # If the file size is <.1 MB we display in KB
+            file_size, size_unit = utils.get_file_size(dist_to_upload)
+            print(f"   {dist_to_upload} ({file_size} {size_unit})")
```
If this is a float, how will this get formatted? What's the accuracy? How many significant digits? We should be consistent here by using format specifiers

VikramJayanthi17 replied:
The output would look like this: 
<img width="922" alt="Screen Shot 2020-06-08 at 2 19 31 PM" src="https://user-images.githubusercontent.com/35483161/84085487-8afb3c80-a99a-11ea-9c07-d5aacbb38117.png">
I determined the number of decimal points to round to(1) by copying how PyPI displays the size of files on their website. The function get_file_size in utils.py determines the unit(KB or MB) returns the size rounded to one decimal place as well as the unit. 

I will add the format specifiers to the string. Additionally, it seems that get_file_size is causing codecov tests to fail. The test I thought of adding to pass the codecov would check that the size of one of the helper dists is accurate and outputs the correct unit. Are there any additional tests you would like to see? Thanks for the help.

callmecampos replied:
W.r.t format specifiers, I think what @sigmavirus24 is saying is that it's probably best not to directly round in the utils function, but to instead print the returned float rounded to a certain number of decimal places using something like the following:

`print(f"   {dist_to_upload} ({file_size:.1f} {size_unit})")`

As an example, `print(f"{2.71828:.2f}")` will automatically round and print `2.72`.

bhrutledge commented on twine/utils.py:
```diff
@@ -159,6 +160,16 @@ def normalize_repository_url(url: str) -> str:
     return urlunparse(parsed)
 
 
+def get_file_size(filepath: str) -> Tuple[float, str]:
+    """Return the file size and unit of size following PyPI's file size format."""
+    size_unit = "MB"
+    file_size = os.path.getsize(filepath) / (1024 * 1024)
+    if round(file_size, 1) == 0:
+        size_unit = "KB"
+        file_size = os.path.getsize(filepath) / 1024
+    return file_size, size_unit
```
Thanks for breaking this out into a utility function. Two nitpicks: I'd use `filename` instead of `filepath` (because the former is more common in the code), and I'm not sure `PyPI's file size format` is as useful as a more explicit description.

Also, I think it could be simplified, e.g.:

```python
    file_size = os.path.getsize(filename) / 1024
    size_unit = "KB"

    if file_size > 1024:
        file_size = file_size / 1024
        size_unit = "MB"

    return file_size, size_unit
```

That said, I think it would be preferable for this to return a string, e.g.:

```python
    return f"{file_size:.1f} {size_unit}"
```

That would simplify the usage in `upload`, and give us a reusable utility for getting a consistently-formatted file size.

bhrutledge commented on tests/test_upload.py:
```diff
@@ -72,6 +73,27 @@ def test_successs_prints_release_urls(upload_settings, stub_repository, capsys):
     assert captured.out.count(NEW_RELEASE_URL) == 1
 
 
+def test_successs_prints_uploaded_package_names_if_verbose(upload_settings, capsys):
+    """Prints the path of each distribution attempting to be uploaded."""
```
Looks like you attempted to follow the naming/docstring conventions in the tests below, which is generally a good idea. However, I think think is more accurate/descriptive, and follows are recent agreement to use imperative mood in docstrings:

```suggestion
def test_print_packages_if_verbose(upload_settings, capsys):
    """Print the path and file size of each distribution attempting to be uploaded."""
```

I fixed the mood in the existing tests in #657.

bhrutledge commented on tests/test_upload.py:
```diff
@@ -72,6 +73,27 @@ def test_successs_prints_release_urls(upload_settings, stub_repository, capsys):
+
+    captured = capsys.readouterr()
+
+    for dist_to_upload in dists_to_upload:
+        assert captured.out.count(dist_to_upload) == 1
```
I think this test should also check for file size.

bhrutledge replied:
I think this would be a good place to check for the fixture sizes that you've currently got in `test_get_file_size`.

bhrutledge commented on tests/test_upload.py:
```diff
@@ -11,6 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
```
Nitpick:

```suggestion
```

bhrutledge commented on twine/commands/upload.py:
```diff
@@ -60,6 +60,12 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
 
     print(f"Uploading distributions to {repository_url}")
+    if upload_settings.verbose:
+        for dist_to_upload in uploads:
+            # If the file size is <.1 MB we display in KB
+            file_size, size_unit = utils.get_file_size(dist_to_upload)
+            print(f"   {dist_to_upload} ({file_size:.1f} {size_unit})")
```
For consistency with the existing loop through `uploads`, let's use `filename` for the variable.

I think the comment would be better as part of the docstring for `get_file_size`.

Also, nitpick: I think two spaces is sufficient for indentation (and follows the `skip_message` indentation later)

```suggestion
        for filename in uploads:
            file_size, size_unit = utils.get_file_size(filename)
            print(f"  {dist_to_upload} ({file_size:.1f} {size_unit})")
```

bhrutledge commented on tests/test_utils.py:
```diff
@@ -313,3 +313,14 @@ def test_check_status_code_for_missing_status_code(capsys, repo_url):
     assert captured.out == "NOTE: Try --verbose to see response content.\n"
+
+
+def test_get_file_size_unit_determination():
+    """Test if get_file_size returns a non zero value in KB for a file < .01 MB."""
+    test_file = helpers.SDIST_FIXTURE
+
+    file_size, size_unit = utils.get_file_size(test_file)
+    print(file_size)
+    print(size_unit)
+    assert file_size != 0
+    assert size_unit == "KB"
```
Nice to see a unit test for `get_file_size`, but could use some work. Specifically, removing the `print` statements, and testing with different file sizes (ideally using [pytest parameterization](https://docs.pytest.org/en/stable/parametrize.html).

Also, it looks like you created this branch before we merged a rewrite of the test docstrings. We're not using words like "Test", opting instead to use imperative mood to describe the expected behavior. See https://github.com/pypa/twine/commit/194eccdde19a2064c3e4a22e7d9364bf9fcf7992 for examples.

VikramJayanthi17 commented:
I added the changes suggested by @bhrutledge, thanks for the help. I wasn't sure what the appropriate docstring would be for the test_get_file_size function, currently it is `"""Compare the size of a file to the expected value."""` but I am open to suggestions on what to change it to.

bhrutledge commented on tests/test_utils.py:
```diff
@@ -313,3 +313,19 @@ def test_check_status_code_for_missing_status_code(capsys, repo_url):
     assert captured.out == "NOTE: Try --verbose to see response content.\n"
+
+
+@pytest.mark.parametrize(
+    ("filename, expected"),
+    [
+        (helpers.WHEEL_FIXTURE, "15.4 KB"),
+        (helpers.SDIST_FIXTURE, "20.8 KB"),
+        (helpers.NEW_SDIST_FIXTURE, "26.1 KB"),
+        (helpers.NEW_WHEEL_FIXTURE, "21.9 KB"),
+    ],
+)
+def test_get_file_size(filename, expected):
+    """Compare the size of a file to the expected value."""
+    file_size = utils.get_file_size(filename)
+
+    assert file_size == expected
```
Nice job with parameterization. However, this doesn't cover the "MB" case ([see the report](https://codecov.io/gh/pypa/twine/compare/89aab15175a89d120802756ed4399d7dd93959bc...3954ad972e062e20f3b132789386ba6d3994c5f8/src/twine/utils.py?before=twine/utils.py)). That said, I'd rather not add a 1 MB file to the repo just to test this.

So, instead using actual files, how about changing the parameters to be a size in bytes and the expected string, then using pytest's [`monkeypatch` fixture](https://docs.pytest.org/en/latest/monkeypatch.html) to make `os.path.getsize` return the parameterized size in bytes?

bhrutledge commented on tests/test_upload.py:
```diff
@@ -72,6 +73,31 @@ def test_successs_prints_release_urls(upload_settings, stub_repository, capsys):
     assert captured.out.count(NEW_RELEASE_URL) == 1
 
 
+def test_successs_prints_uploaded_package_names_if_verbose(upload_settings, capsys):
+    """Print the path and file size of each distribution attempting to be uploaded."""
```
Repeating an [earlier comment](https://github.com/pypa/twine/pull/652#discussion_r437299337): I think think is more accurate/descriptive, and follows are recent agreement to use imperative mood in docstrings:

```suggestion
def test_print_packages_if_verbose(upload_settings, capsys):
    """Print the path and file size of each distribution attempting to be uploaded."""
```

bhrutledge commented on tests/test_utils.py:
```diff
@@ -313,3 +313,19 @@ def test_check_status_code_for_missing_status_code(capsys, repo_url):
+    ],
+)
+def test_get_file_size(filename, expected):
+    """Compare the size of a file to the expected value."""
```
Test docstrings are an opportunity to describe the expected behavior. So, for example:

```suggestion
    """Get the size of file as a string with units."""
```

In this case, it's somewhat redundant with the function docstring, and personally, I would probably skip it for this test, but the project is pushing for more docstrings, so I think it's good to get in the habit.

sigmavirus24 commented on twine/commands/upload.py:
```diff
@@ -60,6 +60,11 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
     repository_url = cast(str, upload_settings.repository_config["repository"])
 
     print(f"Uploading distributions to {repository_url}")
+    if upload_settings.verbose:
+        for filename in uploads:
+            file_size = utils.get_file_size(filename)
+            print(f"  {filename} ({file_size})")
+        print("\n")
```
Breaking that logic apart might be nice @VikramJayanthi17 

I'm starting to question the value in printing the list in advance because it's not as if it's part of a prompt. So when we talk about --verbose we're not talking about increasing noise. We're trying to relatively increase volume of output while keeping a high amount of signal.

di replied:
I think there's value in "logging some indication of what's about to happen right before it happens", in the event that what happens is an error. I don't think it matters too much where this happens, but it should probably be as close to the action as we can make it.

Here, if we want to include what signatures will be uploaded along with the file (which I think would be good), it sounds like we should break the loop apart like @VikramJayanthi17 suggests and log a message right before `repository.upload(package)` that includes everything that's about to happen.

bhrutledge replied:
I like @VikramJayanthi17's proposal as well. I was wondering something similar, but noted that the signing is currently skipped if the package is already uploaded:

https://github.com/pypa/twine/blob/bffcba666875fce1a8e01e3cb991c840e18dd0f9/twine/commands/upload.py#L83-L85

Not sure what the implications of that are.

VikramJayanthi17 replied:
I attempted to address this in the latest [commit](https://github.com/pypa/twine/pull/652/commits/bb971a8deaf5f72dfe97ada0af4fad447d9b1dbf) and outlined my approach in this [comment](https://github.com/pypa/twine/pull/652#issuecomment-642926319).

bhrutledge commented on tests/test_utils.py:
```diff
@@ -313,3 +313,16 @@ def test_check_status_code_for_missing_status_code(capsys, repo_url):
+)
+def test_get_file_size(num_bytes, expected, monkeypatch):
+    """Get the size of file as a string with units."""
+    monkeypatch.setattr(os.path, "getsize", lambda num_bytes: num_bytes)
```
The argument to `getsize` is a filename, but a common pattern (that I think avoids linting errors) is to use an underscore for unused arguments:

```suggestion
    monkeypatch.setattr(os.path, "getsize", lambda _: num_bytes)
```

bhrutledge commented on tests/test_upload.py:
```diff
@@ -72,6 +73,35 @@ def test_successs_prints_release_urls(upload_settings, stub_repository, capsys):
+
+    captured = capsys.readouterr()
+
+    for index in range(len(dists_to_upload)):
+        curr_dist = dists_to_upload[index]
+        size_of_curr_dist = utils.get_file_size(curr_dist)
```
There are more Pythonic ways to loop indexing into a list. One would be to use a dictionary, e.g.:

```python
dists_to_upload = {
    helpers.WHEEL_FIXTURE: "15.4 KB",
    # ...
}

# ...

result = upload.upload(upload_settings, dists_to_upload.keys())

# ...

for filename, size in dists_to_upload.items():
```

Also, I think the call to `get_file_size` isn't necessary, since it's already unit-tested, and this is more of an integration test to make sure that the file size is printed.

Finally, instead of multiple assertions, I might have one that asserts that expected formatted string, i.e. `f"{filename} ({size})"`, is in the output.

bhrutledge commented on tests/test_utils.py:
```diff
@@ -313,3 +313,16 @@ def test_check_status_code_for_missing_status_code(capsys, repo_url):
+    ("num_bytes, expected"),
+    [(3704, "3.6 KB"), (1153433, "1.1 MB"), (21412841, "20.4 MB")],
+)
+def test_get_file_size(num_bytes, expected, monkeypatch):
```
Let's use a more explicit variable name:

```suggestion
def test_get_file_size(size_in_bytes, formatted_size, monkeypatch):
```

VikramJayanthi17 commented:
Added the changes requested by @bhrutledge : https://github.com/pypa/twine/pull/652/commits/eb70d69de29c8fc14199011ad95b0cab0f2de905 

The changes that need to be made regarding [signatures](https://github.com/pypa/twine/pull/652#discussion_r438102370) are still under discussion and I will add any changes to this PR if needed.

VikramJayanthi17 commented:
As discussed [above](https://github.com/pypa/twine/pull/652#discussion_r438102370), I changed the `upload` function by breaking apart the logic of signing and uploading packages. We now print the signature files of the distributions being uploaded. The print statement happens after the `PackageFile` instantiating loop and before the uploads are attempted or the skip messages are printed. This means that we are signing packages that may be skipped as mentioned by @bhrutledge. 
I believe this location for printing would provide the most information to the user(which distributions and signatures we are attempting to upload) without having other messages(like skips) or errors(while uploading) output at the same time. 
<img width="935" alt="Screen Shot 2020-06-11 at 12 46 56 PM" src="https://user-images.githubusercontent.com/35483161/84438225-05220000-abeb-11ea-8208-b5b4a4001889.png">

sigmavirus24 commented on twine/commands/upload.py:
```diff
@@ -58,21 +58,37 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
+        for filename, package in file_package_zip:
             file_size = utils.get_file_size(filename)
             print(f"  {filename} ({file_size})")
+            print(f"  {package.signed_filename} (Signature)")
```
Not ever package will have a signed filename. What happens here if it doesn't have a signature file? What does this print? Folks generating signatures should also know that a `.asc` file is the signature, what are we hoping to communicate by putting `(Signature)` at the end?

VikramJayanthi17 replied:
That makes sense, thanks for the feedback. Would you suggest printing the filename like the other files(without anything to indicate that its a signature)? I will also make this print conditional on upload_settings.sign

bhrutledge commented on twine/commands/upload.py:
```diff
@@ -58,16 +58,37 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
 
     print(f"Uploading distributions to {repository_url}")
+    if upload_settings.verbose:
+        for filename, package in file_package_zip:
+            file_size = utils.get_file_size(filename)
+            print(f"  {filename} ({file_size})")
```
I think you could use `package.filename` instead of `filename`, which I think would remove the need for `file_package_zip` (which, FWIW, I probably would have made a dict like `{filename: package, ...}`.

bhrutledge commented on twine/commands/upload.py:
```diff
@@ -58,16 +58,37 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
     uploads = [i for i in dists if not i.endswith(".asc")]
     upload_settings.check_repository_url()
     repository_url = cast(str, upload_settings.repository_config["repository"])
+    packages_to_upload = []
+
+    for filename in uploads:
```
I really like seeing this behavior extracted from the upload loop, and it gives me more refactoring ideas. My first thought is to pull the whole thing into a separate function, e.g.:

```python
packages_to_upload = _make_packages(uploads, signatures, upload_settings)
```

However, whenever I see this idiom:

```python
new_list = []
for item in other_list:
    new_item = transform(item)
    new_list.append(new_item)
```

I think: can I make this a list comprehension? The answer is usually yes (see: [a good talk](https://www.youtube.com/watch?v=ei71YpmfRX4) from PyCon this year).

In this case, you could use a helper function, e.g.:

```python
packages_to_upload = [
    _make_package(filename, signatures, upload_settings)
    for filename in uploads
]
```

Both of these options approaches would make it easier to unit test.

VikramJayanthi17 commented:
I added the changes @bhrutledge requested. 

> Not ever package will have a signed filename. What happens here if it doesn't have a signature file? What does this print?  

Additionally I added a clause to check that the package signature will only be displayed if upload_settings.sign is true or if the distribution is pre-signed to address this issue that @sigmavirus24 pointed out. Thanks for the help.

bhrutledge commented on twine/commands/upload.py:
```diff
@@ -59,15 +76,23 @@ def upload(upload_settings: settings.Settings, dists: List[str]) -> None:
+    ]
+
     print(f"Uploading distributions to {repository_url}")
+    if upload_settings.verbose:
+        for package in packages_to_upload:
+            file_size = utils.get_file_size(package.filename)
+            print(f"  {package.filename} ({file_size})")
+            if upload_settings.sign or package.signed_filename in signatures:
+                print(f"  {package.signed_filename}")
+        print("\n")
```
Although I didn't state this, I was hoping that the verbosity would end up in the helper function. Also, I think the signature clause isn't accurate; `package.gpg_signature` seems to be the canonical indication of whether or not a package was signed.

In attempt to understand this and make a concrete suggestion, I went down a rabbit hole and implemented this and other changes in the tests. You can see the diff here:

https://github.com/VikramJayanthi17/twine/compare/testing-add-uploaded-packages-to-verbose...bhrutledge:testing-add-uploaded-packages-to-verbose

To expedite the merge of this PR, I'd like to push my changes to your branch. Is that okay? If you like, I can open a separate PR against this branch on your fork, and we can comment on lines there.

VikramJayanthi17 replied:
That makes sense. Pushing the changes to my branch sounds good, thanks for the help!

sigmavirus24 commented on twine/commands/upload.py:
```diff
@@ -50,24 +51,45 @@ def skip_upload(
-
     print(f"Uploading distributions to {repository_url}")
 
+    packages_to_upload = [
```
Do we need to allocate a list here? Would it be better to create a generator and use that in the for-loop? I don't think that's a blocker, just food for thought for @VikramJayanthi17

bhrutledge commented:
@VikramJayanthi17 This was released in 3.2.0. Thanks again for your work on this.
